File: coding/src/com/algos/core/li09_linkedlist/revised/SingleLinkedList.java
package com.algos.core.li09_linkedlist.revised;

import com.algos.core.li30_model.ListNode;

public class SingleLinkedList {

	public ListNode head;
	
	public static void main(String[] args) {
		SingleLinkedList sll = new SingleLinkedList();
		
		sll.print();
		sll.insertAfterEnd(10);
		System.out.println();
		sll.print();
		sll.insertAfterEnd(20);
		sll.insertAfterEnd(30);
		System.out.println();
		sll.print();

		sll.insertBeforeHead(40);
		System.out.println();
		sll.print();
		sll.insertBeforeHead(50);
		System.out.println();
		sll.print();
		
		sll.insertAtKthPosition(60, 1);
		sll.insertAtKthPosition(70, 2);
		sll.insertAtKthPosition(80, 3);
		System.out.println();
		sll.print();
		
		sll.deleteAtKthPosition(1);
		System.out.println();
		sll.print();
		
		sll.reverse();
		System.out.println();
		sll.print();
	}
	
	public void insertAfterEnd(int data){
		if(head == null){
			ListNode node = new ListNode(data);
			head = node;
			return;
		}
		ListNode temp = head;
		while(temp.next!= null){
			temp = temp.next;
		}
		
		temp.next = new ListNode(data);
		
	}
	public void insertBeforeHead(int data){
		ListNode node = new ListNode(data);
		node.next = head;
		head = node;
	}
	public void print(){
		if(head == null){
			System.out.println("List is empty");
			return;
		}
		
		ListNode temp = head;
		while(temp != null){
			System.out.print(temp.val+" ");
			temp = temp.next;
		}
	}
	public void insertAtKthPosition(int data, int k){
		ListNode node  = new ListNode(data);
		if(k == 1){
			node.next = head;
			head = node;
			return;
		}
		ListNode temp = head;
		for(int i=0; i<k-2; i++){
			temp = temp.next;
		}
		node.next = temp.next;
		temp.next = node;
	}
	
	public void deleteAtKthPosition(int k){
		ListNode temp = head;
		if(k == 1){
			head = head.next;
			return;
		}
		for(int i=0; i<k-2; i++){
			temp = temp.next;
		}

		temp.next = temp.next.next;
	}
	
	public void reverse(){
		ListNode prev = null, curr = null, next;
		curr = head;
		while(curr != null){
			next = curr.next;
			curr.next = prev;
			prev = curr;
			curr = next;
		}
		head = prev;
	}
	
	public void printRec(ListNode node){
		if(node == null){
			return;
		}
		System.out.print(node.val+" ");
		printRec(node.next);
	}
	
	public void reversePrintRec(ListNode node){
		if(node == null){
			return;
		}
		reversePrintRec(node.next);
		System.out.print(node.val+" ");
	}

	public int size() {
		int count = 0;
		ListNode curr = head;
		while(curr != null) {
			curr = curr.next;
			count++;
		}
		return count;
	}
	
}

File: dsa/topics/09_linkedlist/python/single_linked_list.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class SingleLinkedList:

    def __init__(self):
        self.head = None

    def insertAfterEnd(self, data):
        if self.head is None:
            self.head = ListNode(data)
            return

        temp = self.head
        while temp.next:
            temp = temp.next
        temp.next = ListNode(data)

    def insertBeforeHead(self, data):
        node = ListNode(data)
        node.next = self.head
        self.head = node

    def print(self):
        if self.head is None:
            print("List is empty")
            return

        temp = self.head
        while temp:
            print(temp.val, end=" ")
            temp = temp.next
        print()

    def insertAtKthPosition(self, data, k):
        node = ListNode(data)

        if k == 1:
            node.next = self.head
            self.head = node
            return

        temp = self.head
        for _ in range(k - 2):
            temp = temp.next

        node.next = temp.next
        temp.next = node

    def deleteAtKthPosition(self, k):
        if self.head is None:
            return

        if k == 1:
            self.head = self.head.next
            return

        temp = self.head
        for _ in range(k - 2):
            temp = temp.next

        temp.next = temp.next.next

    def reverse(self):
        prev = None
        curr = self.head

        while curr:
            next_node = curr.next
            curr.next = prev
            prev = curr
            curr = next_node

        self.head = prev

    def printRec(self, node):
        if node is None:
            return
        print(node.val, end=" ")
        self.printRec(node.next)

    def reversePrintRec(self, node):
        if node is None:
            return
        self.reversePrintRec(node.next)
        print(node.val, end=" ")

    def size(self):
        count = 0
        curr = self.head
        while curr:
            count += 1
            curr = curr.next
        return count


# Demo (same flow as Java main)
if __name__ == "__main__":
    sll = SingleLinkedList()

    sll.print()
    sll.insertAfterEnd(10)
    sll.print()

    sll.insertAfterEnd(20)
    sll.insertAfterEnd(30)
    sll.print()

    sll.insertBeforeHead(40)
    sll.print()

    sll.insertBeforeHead(50)
    sll.print()

    sll.insertAtKthPosition(60, 1)
    sll.insertAtKthPosition(70, 2)
    sll.insertAtKthPosition(80, 3)
    sll.print()

    sll.deleteAtKthPosition(1)
    sll.print()

    sll.reverse()
    sll.print()
